[COLAB-2013] Logo Carousel Widget

umov opened the pull request:
This PR implements the Logo Carousel Widget (COLAB-2013).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cci-mit/xcolab/18)
<!-- Reviewable:end -->

jobachh commented:
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 24 unresolved discussions, some commit checks failed.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java, line 19 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsZHRwfQCiwy4lDpe0i:-KsZHRwfQCiwy4lDpe0j:b-z7mf5p) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java#L19)):*
> ```Java
> public class CarouselController {
> 
>     private static final Logger _log = LoggerFactory.getLogger(CarouselController.class);
> ```

This logger seems to be unused - you can remove it.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java, line 21 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsZCtC9VL1yl-UK1vNy:-KsZCtC9VL1yl-UK1vNz:b-11uw3d) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java#L21)):*
> ```Java
>     private static final Logger _log = LoggerFactory.getLogger(CarouselController.class);
> 
>     public CarouselController() {
> ```

The empty constructor is automatically generated when you don't specify any constructor.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java, line 25 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsZD6v6Jp53XJTJh_ff:-KsZD6v6Jp53XJTJh_fg:boh83my) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java#L25)):*
> ```Java
> 
>     @GetMapping("/carouselwidget")
>     public String showCarousel(@RequestParam(required = false) String preferenceId, HttpServletRequest request, HttpServletResponse response, Model model)  {
> ```

Those handler methods can get a lot of parameters, so we try to keep the order consistent - the  request, response, and model should come before the `@RequestParam`. You can find the convention in our [code style guide](https://github.com/CCI-MIT/XCoLab/wiki/Code-Style#conventions).

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java, line 27 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsZDt7hva2bKrHlhNle:-KsZDt7hva2bKrHlhNlf:b-z0zp8q) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java#L27)):*
> ```Java
>     public String showCarousel(@RequestParam(required = false) String preferenceId, HttpServletRequest request, HttpServletResponse response, Model model)  {
>         Locale locale = LocaleContextHolder.getLocale();
>         CarouselPreferences carouselPreferences = new CarouselPreferences(preferenceId,locale.getLanguage());
> ```

There should be a space after the comma.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java, line 14 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-Ksa9GH45MLSJ0Ohbg3i:-Ksa9GH45MLSJ0Ohbg3j:bagbeg0) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java#L14)):*
> ```Java
> 
> import java.io.IOException;
> import java.util.ArrayList;
> ```

This import is unused.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java, line 25 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaFE6xllmZxd9jSnqP:-KsaFE6xllmZxd9jSnqQ:b-bcrjxd) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java#L25)):*
> ```Java
>     private String title;
>     private List<LogoElement> logos;
>     private Integer logosCount;
> ```

We generally try to stick to the primitive number types (`int`) unless we explicitly need the boxed object (e.g. to store it in a collection). This can make things such as equality comparisons simpler down the line and makes it clear that the field is not nullable.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java, line 56 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsZHAHxljYOZVRfZgUR:-KsZHAHxljYOZVRfZgUS:b-787vg) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java#L56)):*
> ```Java
>     }
> 
>     public CarouselPreferences() {
> ```

Constructors should come right after the field definitions, before any other methods. I just added a [section about source file structure](https://github.com/CCI-MIT/XCoLab/wiki/Code-Style#additional-rules) to our code style guide.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java, line 70 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaHY8orX_9lp82ldC7:-KsaHY8orX_9lp82ldC8:boetpv5) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java#L70)):*
> ```Java
>         if (prefs.has(LOGOS_PREFERENCE)) {
>             JSONArray logosArray = prefs.getJSONArray(LOGOS_PREFERENCE);
>             ObjectMapper mapper = new ObjectMapper();
> ```

Object mappers are thread safe, so you can safely turn this into a `static final` field. Since we are only reading, you can go one further use the immutable `ObjectReader` class:
```java
private static final ObjectReader logoListReader = new ObjectMapper().readerFor(new TypeReference<List<LogoElement>>() {});
```

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java, line 73 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaG9G_vE1xIz7xUi48:-KsaG9GayhMvZf2wmQxN:blzvcf7) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java#L73)):*
> ```Java
>             try {
>                 logos = mapper.readValue(logosArray.toString(), new TypeReference<List<LogoElement>>() {});
>             } catch (IOException ignored) {
> ```

This exception should not occur under normal circumstances, but it's best not to hide it. The simplest way is to re-throw this checked `Exception` as a subclass of`RuntimeException`.  I think here an `IllegalStateException` would make sense, otherwise we also have a custom `InternalException` if no existing exception fits. Make sure you pass the cause in the constructor to help debugging should it ever be thrown.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferencesController.java, line 39 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-Ks_CVDOeRscqvnFAAua:-Ks_CVDOeRscqvnFAAub:b-n19ql7) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferencesController.java#L39)):*
> ```Java
> 
>         CarouselPreferences carouselPreferences = new CarouselPreferences(preferenceId, language);
>         carouselPreferences.setLogosCount(carouselPreferences.getLogos().size());
> ```

This line throws an NPE when the preferences are uninitialized (as `carouselPreferences.getLogos()` returns `null`), it should probably default to `0` then.

Also: Could we move this to the constructor of `CarouselPreferences`?

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/LogoElement.java, line 48 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-Ks_CwN1WJ_2jXAgruKo:-Ks_CwN1WJ_2jXAgruKp:bca5doy) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/LogoElement.java#L48)):*
> ```Java
>     @JsonCreator
>     public LogoElement(@JsonProperty("imageUrl") String imageUrl,
>             @JsonProperty("linkUrl") String linkUrl, @JsonProperty("altText") String altText) {
> ```

Are these annotations required? As long as the setters and getters follow the bean convention (i.e. have `getX` and `setX` style getters and setters), jackson should be able to do this without annotations as long as there is an empty default constructor (as the one below).

Also: note the [source file structure](https://github.com/CCI-MIT/XCoLab/wiki/Code-Style#additional-rules)  mentioned in another comment (constructor should come right after field definitions).

---

*[view/src/main/resources/static/sass/partials/components/_c-CarouselWrapper.scss, line 13 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-Ks_DUzIycX1l4uhQEkc:-Ks_DUzIycX1l4uhQEkd:btgzwrv) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/resources/static/sass/partials/components/_c-CarouselWrapper.scss#L13)):*
> ```SCSS
>   overflow: hidden;
>   text-align: center;
>   margin-bottom: $carousel-heading-margin-bottom !important;
> ```

Do we need `!important` here? I think the rule is specific enough to override the default h2 rule.

---

*[view/src/main/resources/static/sass/partials/components/_c-CarouselWrapper.scss, line 19 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaJUsQOBKdAeP42hIZ:-KsaJUsQOBKdAeP42hI_:b-tc91r) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/resources/static/sass/partials/components/_c-CarouselWrapper.scss#L19)):*
> ```SCSS
> 
> .c-CarouselWrapper h2:before,
> .c-CarouselWrapper h2:after {
> ```

Since we are using SCSS, you can nest these rules inside the containing element:

```scss
.c-CarouselWrapper {

  // rules for c-CarouselWrapper

  h2 {

    // rules for .c-CarouselWrapper h2

    &:before, &:after {
      // ...
    }
    
    &:before {
      // ...
    }

    &:after {
      // ...
    }

}
```

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 10 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaM82UFVpLScaybCbS:-KsaM82UFVpLScaybCbT:bqzfczl) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L10)):*
> ```
>     <xcolab:layout>
>         <form:form action="/carouselwidget/savePreferences" commandName="carouselPreferences">
>             <form:hidden path="logosCount" id="logosCount" value="${carouselPreferences.logos.size()}" />
> ```

IntelliJ will likely show you a warning that this variable isn't defined (and as a result it won't provide you with any code completion or smart navigation). You can solve that by defining your variables (usually at the top of the file/scope):
```jsp
<!--@elvariable id="carouselPreferences" type="org.xcolab.view.pages.carouselwidget.CarouselPreferences"-->
```

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 33 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaC9XC8esse8tTXee0:-KsaC9XC8esse8tTXee1:b-e5mgiz) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L33)):*
> ```
>                         </select>
> 
>                         <form:hidden path="preferenceId" id="newPref"  cssStyle="margin-left:20px;"/>
> ```

The `cssStyle` attribute has no effect here since the input is hidden.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 35 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaMZ83YPPq7il63EP9:-KsaMZ83YPPq7il63EPA:bus2snz) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L35)):*
> ```
>                         <form:hidden path="preferenceId" id="newPref"  cssStyle="margin-left:20px;"/>
> 
>                         <script>
> ```

Let's move this script block to the bottom of the file to keep markup and JavaScript separate.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 43 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaAJZsK5D755QjbDi2:-KsaAJZsK5D755QjbDi3:bon9fl1) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L43)):*
> ```
>                                         $("#newPref").attr("type","text");
>                                         $("#newPref").attr("placeholder","preference identification");
>                                     }else{
> ```

For JavaScript code style, we try to follow the same guidelines as for Java code (where it makes sense). For this section, I'm referring to the [horizontal spacing](https://github.com/CCI-MIT/XCoLab/wiki/Code-Style#horizontal-whitespace) after commas and around keywords and braces/parentheses.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 86 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaLbYXlN8_-9NBmocs:-KsaLbYXlN8_-9NBmoct:b-51w8vj) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L86)):*
> ```
>                     </td>
>                     <td style="width:70%;padding-right:10px;">
>                         <form:input path="title" style="width:100%;"/>
> ```

As a simple way to make inputs and selects look nicer (and avoid the `style` attribute) you can use the bootstrap class `form-control`.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 91 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaEqPXx-ZimGe-YINw:-KsaEqPXx-ZimGe-YINx:b-1i40fi) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L91)):*
> ```
>             </table>
> 
>             <table class="table" id="logotable">
> ```

This id doesn't seem to be used anywhere, you can remove it.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 102 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaBipfBntNGSYk-Sqa:-KsaBipfBntNGSYk-Sqb:bsvaj6e) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L102)):*
> ```
>                         <td><form:input path="logos[${status.index}].imageUrl" name="imageUrl" id="imageUrl" style="width:100%;"/></td>
>                         <td><form:input path="logos[${status.index}].linkUrl" name="linkUrl" id="linkUrl" style="width:100%;"/></td>
>                         <td><form:input path="logos[${status.index}].altText" name="altText" id="altText" style="width:100%;"/></td>
> ```

Same as above, you can use `class="form-control"` on these inputs.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 114 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaCl-PdKS0V69rKAgI:-KsaCl-PdKS0V69rKAgJ:bx6w5ed) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L114)):*
> table-add
To make it clear that a CSS class is only used to identify an element for jQuery and to make it easier to figure out what it's for, you can use the `js` prefix followed by the component and optionally a specifier, e.g. `js-Table__add`.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 118 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaLAZjfiR1WQgkDg9J:-KsaLAZjfiR1WQgkDg9K:bm38xdq) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L118)):*
> prototype-row
Same as above, you could call it`js-Table__prototypeRow`.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 123 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaDydy00tjQNi15BBX:-KsaDydy00tjQNi15BBY:bhghv1b) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L123)):*
> table-remove
Same as above, `js-Table__remove`.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/showCarousel.jspx, line 17 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaA0FseZcNhNSFRKDx:-KsaA0FseZcNhNSFRKDy:b9ttua8) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/showCarousel.jspx#L17)):*
> ```
>                     <c:choose>
>                         <c:when test="${logo.linkUrl != null and !logo.linkUrl.isEmpty()}">
>                             <a href="${logo.linkUrl}">
> ```

Let's have these open in a new tab by default.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cci-mit/xcolab/18)*
<!-- Sent from Reviewable.io -->

jobachh commented:
@umov Great job overall, I tried it out and it's working very well 👍 I left some notes and small improvements on reviewable, let me know when you've had a chance to look over them.

jobachh commented:
Review status: all files reviewed at latest revision, 25 unresolved discussions, some commit checks failed.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java, line 53 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaP2Vy1Uewut6K2eVW:-KsaP2Vy1Uewut6K2eVX:b-99f69f) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java#L53)):*
> PORTLET_CONTESTS_PREFERENCES
You forgot to change this, it should be `PORTLET_CAROUSEL_PREFERENCES`!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cci-mit/xcolab/18)*
<!-- Sent from Reviewable.io -->

umov commented:
Review status: 1 of 10 files reviewed at latest revision, 25 unresolved discussions.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java, line 19 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsZHRwfQCiwy4lDpe0i:-KseRH4awjrumal-ffr2:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java#L19)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

This logger seems to be unused - you can remove it.
</blockquote></details>

Done.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java, line 21 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsZCtC9VL1yl-UK1vNy:-KseRJql7gN5BUPgS80n:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java#L21)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

The empty constructor is automatically generated when you don't specify any constructor.
</blockquote></details>

Done.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java, line 25 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsZD6v6Jp53XJTJh_ff:-KseRKnEd4NSP_hW0-hF:ble1aq) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java#L25)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

Those handler methods can get a lot of parameters, so we try to keep the order consistent - the  request, response, and model should come before the `@RequestParam`. You can find the convention in our [code style guide](https://github.com/CCI-MIT/XCoLab/wiki/Code-Style#conventions).
</blockquote></details>

Done. FYI: The order was taken from `showContests` in ContestsController.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java, line 27 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsZDt7hva2bKrHlhNle:-KseRTmv4n9LL0BWrP0j:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java#L27)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

There should be a space after the comma.
</blockquote></details>

Done.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java, line 14 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-Ksa9GH45MLSJ0Ohbg3i:-KseUUJke0JRaea9f4EV:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java#L14)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

This import is unused.
</blockquote></details>

Done.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java, line 25 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaFE6xllmZxd9jSnqP:-KseUUnowKbwA3Efj3Lo:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java#L25)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

We generally try to stick to the primitive number types (`int`) unless we explicitly need the boxed object (e.g. to store it in a collection). This can make things such as equality comparisons simpler down the line and makes it clear that the field is not nullable.
</blockquote></details>

Done.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java, line 53 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaP2Vy1Uewut6K2eVW:-KseUXpnF3WR_d8CbO9f:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java#L53)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

> PORTLET_CONTESTS_PREFERENCES
You forgot to change this, it should be `PORTLET_CAROUSEL_PREFERENCES`!
</blockquote></details>

Done.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java, line 56 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsZHAHxljYOZVRfZgUR:-KseUYU_PB3HC4yjcdTE:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java#L56)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

Constructors should come right after the field definitions, before any other methods. I just added a [section about source file structure](https://github.com/CCI-MIT/XCoLab/wiki/Code-Style#additional-rules) to our code style guide.
</blockquote></details>

Done.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java, line 70 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaHY8orX_9lp82ldC7:-KseUZ23DkKwSoYW0ekP:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java#L70)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

Object mappers are thread safe, so you can safely turn this into a `static final` field. Since we are only reading, you can go one further use the immutable `ObjectReader` class:
```java
private static final ObjectReader logoListReader = new ObjectMapper().readerFor(new TypeReference<List<LogoElement>>() {});
```
</blockquote></details>

Done.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java, line 73 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaG9G_vE1xIz7xUi48:-KseUZVZo-xud9I33ewc:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferences.java#L73)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

This exception should not occur under normal circumstances, but it's best not to hide it. The simplest way is to re-throw this checked `Exception` as a subclass of`RuntimeException`.  I think here an `IllegalStateException` would make sense, otherwise we also have a custom `InternalException` if no existing exception fits. Make sure you pass the cause in the constructor to help debugging should it ever be thrown.
</blockquote></details>

Done.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferencesController.java, line 39 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-Ks_CVDOeRscqvnFAAua:-KseU_7HKt5lIqRvwzCG:bgjkuzs) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselPreferencesController.java#L39)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

This line throws an NPE when the preferences are uninitialized (as `carouselPreferences.getLogos()` returns `null`), it should probably default to `0` then.

Also: Could we move this to the constructor of `CarouselPreferences`?
</blockquote></details>

Done by moving it to the constructor of `CarouselPreferences`.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/LogoElement.java, line 48 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-Ks_CwN1WJ_2jXAgruKo:-KsjBMqtdOvc1kQOMfiH:b-cjzfss) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/LogoElement.java#L48)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

Are these annotations required? As long as the setters and getters follow the bean convention (i.e. have `getX` and `setX` style getters and setters), jackson should be able to do this without annotations as long as there is an empty default constructor (as the one below).

Also: note the [source file structure](https://github.com/CCI-MIT/XCoLab/wiki/Code-Style#additional-rules)  mentioned in another comment (constructor should come right after field definitions).
</blockquote></details>

Done. Removed the annotated constructor.

---

*[view/src/main/resources/static/sass/partials/components/_c-CarouselWrapper.scss, line 13 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-Ks_DUzIycX1l4uhQEkc:-KseKNNvkA0vZaz2e62B:b-k35vu1) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/resources/static/sass/partials/components/_c-CarouselWrapper.scss#L13)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

Do we need `!important` here? I think the rule is specific enough to override the default h2 rule.
</blockquote></details>

Yes, we do. `.home #content h2` has an ID in there and overrides this rule.

---

*[view/src/main/resources/static/sass/partials/components/_c-CarouselWrapper.scss, line 19 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaJUsQOBKdAeP42hIZ:-KsjBQfQF1hDmEehPT30:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/resources/static/sass/partials/components/_c-CarouselWrapper.scss#L19)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

Since we are using SCSS, you can nest these rules inside the containing element:

```scss
.c-CarouselWrapper {

  // rules for c-CarouselWrapper

  h2 {

    // rules for .c-CarouselWrapper h2

    &:before, &:after {
      // ...
    }
    
    &:before {
      // ...
    }

    &:after {
      // ...
    }

}
```
</blockquote></details>

Done.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 10 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaM82UFVpLScaybCbS:-KsjEKIbGEK-rDZCy84T:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L10)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

IntelliJ will likely show you a warning that this variable isn't defined (and as a result it won't provide you with any code completion or smart navigation). You can solve that by defining your variables (usually at the top of the file/scope):
```jsp
<!--@elvariable id="carouselPreferences" type="org.xcolab.view.pages.carouselwidget.CarouselPreferences"-->
```
</blockquote></details>

Done.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 33 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaC9XC8esse8tTXee0:-KsjEKqXHfognT9EwLXD:bymuaaa) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L33)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

The `cssStyle` attribute has no effect here since the input is hidden.
</blockquote></details>

Not entirely true. Yes, the input is hidden, but it's revealed when the user selects CREATE NEW PREFERENCE. Changed to styling by `form-control` class.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 35 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaMZ83YPPq7il63EP9:-KsjES5ZhDrV9N1FQi8c:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L35)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

Let's move this script block to the bottom of the file to keep markup and JavaScript separate.
</blockquote></details>

Done.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 43 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaAJZsK5D755QjbDi2:-KsjFRLq98zVpKfFns1C:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L43)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

For JavaScript code style, we try to follow the same guidelines as for Java code (where it makes sense). For this section, I'm referring to the [horizontal spacing](https://github.com/CCI-MIT/XCoLab/wiki/Code-Style#horizontal-whitespace) after commas and around keywords and braces/parentheses.
</blockquote></details>

Done.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 86 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaLbYXlN8_-9NBmocs:-KsjJfp5Y4vlFNz6PVzs:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L86)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

As a simple way to make inputs and selects look nicer (and avoid the `style` attribute) you can use the bootstrap class `form-control`.
</blockquote></details>

Done.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 91 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaEqPXx-ZimGe-YINw:-KsjJltyMOPgyB2cVneJ:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L91)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

This id doesn't seem to be used anywhere, you can remove it.
</blockquote></details>

Done.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 102 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaBipfBntNGSYk-Sqa:-KsjJmlT5oM8J3SzH2IB:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L102)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

Same as above, you can use `class="form-control"` on these inputs.
</blockquote></details>

Done.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 114 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaCl-PdKS0V69rKAgI:-KsjKLVDmo4q3klyZ20Q:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L114)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

> table-add
To make it clear that a CSS class is only used to identify an element for jQuery and to make it easier to figure out what it's for, you can use the `js` prefix followed by the component and optionally a specifier, e.g. `js-Table__add`.
</blockquote></details>

Done.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 118 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaLAZjfiR1WQgkDg9J:-KsjKSrUOl1yVGe7oi2o:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L118)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

> prototype-row
Same as above, you could call it`js-Table__prototypeRow`.
</blockquote></details>

Done.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 123 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaDydy00tjQNi15BBX:-KsjKTfvbEvEaXaBfD3m:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L123)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

> table-remove
Same as above, `js-Table__remove`.
</blockquote></details>

Done.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/showCarousel.jspx, line 17 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaA0FseZcNhNSFRKDx:-KsjLf5Qvoqa5RLjBMz9:b-896fix) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/showCarousel.jspx#L17)):*
<details><summary><i>Previously, jobachh (Johannes Bachhuber) wrote…</i></summary><blockquote>

Let's have these open in a new tab by default.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cci-mit/xcolab/18)*
<!-- Sent from Reviewable.io -->

umov commented:
@jobachh Thanks for the review. Adapted the implementation according to your feedback. Please let me know what you think of the updated version.

jobachh commented:
Reviewed 9 of 9 files at r2.
Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java, line 25 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsZD6v6Jp53XJTJh_ff:-KskFylTgepOINdTfj5i:b-uqha73) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/java/org/xcolab/view/pages/carouselwidget/CarouselController.java#L25)):*
<details><summary><i>Previously, umov wrote…</i></summary><blockquote>

Done. FYI: The order was taken from `showContests` in ContestsController.
</blockquote></details>

Thanks for pointing that out - I adjusted the code style in the `ContestsController` accordingly.

---

*[view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx, line 33 at r1](https://reviewable.io:443/reviews/cci-mit/xcolab/18#-KsaC9XC8esse8tTXee0:-Ksk7cA-M_elqvu0X52f:btmn158) ([raw file](https://github.com/cci-mit/xcolab/blob/388c8f726521cdafe9767ed864f848a5e044e2f2/view/src/main/webapp/WEB-INF/jsp/carouselwidget/editPreferences.jspx#L33)):*
<details><summary><i>Previously, umov wrote…</i></summary><blockquote>

Not entirely true. Yes, the input is hidden, but it's revealed when the user selects CREATE NEW PREFERENCE. Changed to styling by `form-control` class.
</blockquote></details>

I see, in that case the style makes sense - let's keep it like this for now (as it also works like this in the other preferences).

In a separate issue to improve the preferences in general, we can update this to be hidden with css rather than the input type, as that's less surprising to the reader.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cci-mit/xcolab/18)*
<!-- Sent from Reviewable.io -->

jobachh commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cci-mit/xcolab/18)*
<!-- Sent from Reviewable.io -->
